libqtile/confreader.py

```diff
@@ -118,43 +120,118 @@ def _reload_config_submodules(self, path: Path) -> None:
                 if folder in subpath.parents:
                     importlib.reload(module)
 
-    def load(self):
-        if not self.file_path:
-            return
+    def load(self) -> None:
+        new_config = self._do_load()
+        self.__dict__.update(new_config)
 
-        path = Path(self.file_path)
-        name = path.stem
-        sys.path.insert(0, path.parent.as_posix())
+    def validate(self) -> None:
+        self._do_load()
 
-        if name in sys.modules:
-            self._reload_config_submodules(path)
-            config = importlib.reload(sys.modules[name])
-        else:
-            config = importlib.import_module(name)
+    def _do_load(self) -> dict[str, Any]:
+        path = self._file_path
 
-        self.update(**vars(config))
+        if path is None:
+            # This is a test config that doesn't load from file. Instead, config values
+            # were specified as class attributes on a subclass.
+            return {k: v for k, v in vars(self.__class__).items() if not k.startswith("_")}
 
-    def validate(self) -> None:
+        name = path.stem
+
+        try:
+            if name in sys.modules:
+                self._reload_config_submodules(path)
+                config = importlib.reload(sys.modules[name])
+            else:
+                config = importlib.import_module(name)
+        except Exception as e:
+            raise ConfigError from e
+
+        settings = vars(config)
+        available_keys = self._types.keys()
+        specified_keys = available_keys & settings.keys()
+        new_config = self._defaults.copy()
+
+        # All values must pass type checking otherwise none are updated.
+        errors = set()
+        for key in specified_keys:
+            valid_type = self._types[key]
+            value = settings[key]
+
+            if self._check_type(valid_type, value):
+                new_config[key] = value
+            else:
+                errors.add(key)
+
+        if errors:
+            keys = ", ".join(errors)
+            raise ConfigError(f"Invalid types: {keys}")
```

Comment:
Are we sure our typing is robust enough for us to raise a `ConfigError` on an invalid type? Great if the answer is yes!

Reply:
I believe so. Some config options are typed as `Any` so will always pass the validation test - these are the ones where typing isn't fully set up in the most precise way, but that isn't really an issue, except that e.g. `dgroups_key_binder = "not a function"` will get loaded but then later used as a function and raise an error at runtime that won't appear to be related with the users config. For a future PR.